expression/builtin_compare.go

```diff
@@ -82,40 +88,208 @@ var (
+		}
+	}
+
+	switch retEvalTp {
+	case tpInt:
+		sig = &builtinCoalesceIntSig{baseIntBuiltinFunc{bf}}
+	case tpReal:
+		sig = &builtinCoalesceRealSig{baseRealBuiltinFunc{bf}}
+	case tpDecimal:
+		sig = &builtinCoalesceDecimalSig{baseDecimalBuiltinFunc{bf}}
+	case tpString:
+		sig = &builtinCoalesceStringSig{baseStringBuiltinFunc{bf}}
+	case tpTime:
+		sig = &builtinCoalesceTimeSig{baseTimeBuiltinFunc{bf}}
+	case tpDuration:
+		sig = &builtinCoalesceDurationSig{baseDurationBuiltinFunc{bf}}
+	}
+
 	return sig.setSelf(sig), nil
 }
 
-type builtinCoalesceSig struct {
-	baseBuiltinFunc
+// See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
+type builtinCoalesceIntSig struct {
+	baseIntBuiltinFunc
 }
 
-func (b *builtinCoalesceSig) eval(row []types.Datum) (types.Datum, error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinCoalesceIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for _, a := range b.getArgs() {
+		arg, isNull, err := a.EvalInt(row, sc)
+		if err != nil {
+			return arg, isNull, errors.Trace(err)
+		}
+		if !isNull {
+			return arg, false, nil
+		}
 	}
-	return builtinCoalesce(args, b.ctx)
+	return 0, true, nil
+}
+
+// See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
+type builtinCoalesceRealSig struct {
+	baseRealBuiltinFunc
+}
+
+func (b *builtinCoalesceRealSig) evalReal(row []types.Datum) (float64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for _, a := range b.getArgs() {
+		arg, isNull, err := a.EvalReal(row, sc)
+		if err != nil {
+			return arg, isNull, errors.Trace(err)
+		}
+		if !isNull {
+			return arg, false, nil
+		}
+	}
+	return 0, true, nil
 }
 
-// builtinCoalesce returns the first non-NULL value in the list,
-// or NULL if there are no non-NULL values.
 // See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
-func builtinCoalesce(args []types.Datum, ctx context.Context) (d types.Datum, err error) {
-	for _, d = range args {
-		if !d.IsNull() {
-			return d, nil
+type builtinCoalesceDecimalSig struct {
+	baseDecimalBuiltinFunc
+}
+
+func (b *builtinCoalesceDecimalSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for _, a := range b.getArgs() {
+		arg, isNull, err := a.EvalDecimal(row, sc)
+		if err != nil {
+			return arg, isNull, errors.Trace(err)
+		}
+		if !isNull {
+			return arg, false, nil
+		}
+	}
+	return nil, true, nil
+}
+
+// See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
+type builtinCoalesceStringSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinCoalesceStringSig) evalString(row []types.Datum) (string, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for _, a := range b.getArgs() {
+		arg, isNull, err := a.EvalString(row, sc)
+		if err != nil {
```

Comment:
can we use the following code, seems to be more elegant.
``` go
func (b *builtinCoalesceStringSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
    for _, a := range b.getArgs(){
        if err !=nil || !isNull {
            break
        }
    }
    return res, isNull, errors.Trace(err)
}
```

so as the other 5 functions.